pepper-apis/src/main/java/org/broadinstitute/ddp/filter/StudyLanguageResolutionFilter.java

```diff
@@ -0,0 +1,98 @@
+            Matcher matcher = Pattern.compile(STUDY_GUID_REGEX).matcher(request.url());
+            String studyGuid = matcher.find() ? matcher.group(STUDY_GUID_INDEX) : null;
+            // The "supported languages" notion is an attribute of a study, thus is doesn't
+            // make any sense outside of the study context
+            boolean supportedLanguagesCanBeDetected = studyGuid != null;
+            if (!supportedLanguagesCanBeDetected) {
+                LOG.warn(
+                        "Supported languages can't be detected because the filter is invoked"
+                        + " before the route that is outside of the study context. Please"
+                        + " remount the filter under '/user/*/studies/*' instead. Current path = {}",
+                        request.url()
+                );
+                return;
+            }
+            LOG.info("The supported languages can be detected for the study {}", studyGuid);
+            Locale ddpAuthPreferredLocale = RouteUtil.getDDPAuth(request).getPreferredLocale();
+            LanguageDto preferredLanguage = TransactionWrapper.withTxn(
+                    handle -> getPreferredLanguage(handle, acceptLanguageHeader, ddpAuthPreferredLocale, studyGuid)
+            );
+            request.attribute(USER_LANGUAGE, preferredLanguage);
+            LOG.info("Added the preferred user language {} to the attribute store", preferredLanguage.getIsoCode());
+        } catch (Exception e) {
+            LOG.error("Error while figuring out the user language", e);
+        }
+    }
+
+    public LanguageDto getPreferredLanguage(Handle handle, String acceptLanguageHeader, Locale ddpAuthPreferredLocale, String studyGuid) {
+        List<LanguageRange> acceptLanguages = StringUtils.isNotEmpty(acceptLanguageHeader)
+                ? LanguageRange.parse(acceptLanguageHeader) : Collections.emptyList();
+        Set<LanguageDto> languagesSupportedByStudy = handle.attach(StudyDao.class).findSupportedLanguagesByGuid(studyGuid);
+        Map<Locale, LanguageDto> supportedStudyLocaleToLang = languagesSupportedByStudy.stream()
+                .collect(Collectors.toMap(LanguageDto::toLocale, lang -> lang));
+        Locale preferredLocale = I18nUtil.resolvePreferredLanguage(
+                ddpAuthPreferredLocale, acceptLanguages, supportedStudyLocaleToLang.keySet()
+        );
+        // If we are lucky, the preferred language will be among the list
+        // of supported languages and we'll spare a db call. Otherwise, we
+        // will have to look into the db to get the id of the fallback language
+        LanguageDto preferredLanguage = Optional.ofNullable(supportedStudyLocaleToLang.get(preferredLocale))
+                .orElseGet(
+                        () -> new LanguageDto(
+                                handle.attach(JdbiLanguageCode.class).getLanguageCodeId(preferredLocale.getLanguage()),
```

Comment:
Now that we have `LanguageDto`, I'll recommend adding a `findLanguage() -> LanguageDto` method instead of `new`-ing one up.